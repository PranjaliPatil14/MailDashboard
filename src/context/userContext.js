import React, { useEffect, useState } from "react";
import * as PropTypes from "prop-types";
import useLocalStorage, { USER } from "../utilities/useLocalStorage";

const initialUserContextValue = {
  users: [],
  loggedInUser: {
    id: 0,
    name: "",
    role: "",
    profileImage: "",
  },
  setLoggedInUser: () => {},
  setUsers: () => {},
};

export const UserContext = React.createContext(initialUserContextValue);

const UserContextProvider = ({ children }) => {
  const [users, setUsers] = useState([]);
  const [loggedInUser, setLoggedInUser] = useState({});
  const { value } = useLocalStorage(USER);

  useEffect(() => {
    if (!users.length) {
      setUsers(value);
    }
  }, [users.length, value]);

  return (
    <UserContext.Provider
      value={{ users, setUsers, loggedInUser, setLoggedInUser }}
    >
      {children}
    </UserContext.Provider>
  );
};

UserContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default UserContextProvider;
