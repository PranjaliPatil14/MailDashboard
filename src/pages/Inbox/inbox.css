.page {
    background: #F3F3F4;
    flex: 1;
}

.page-content {
    padding: 20px 20px 0 30px;
    display: flex;
    gap: 30px;
}
