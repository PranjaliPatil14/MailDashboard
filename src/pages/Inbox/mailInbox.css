.page {
    background: #F3F3F4;
    flex: 1;
}

.page-content {
    padding: 20px 30px 0 30px;
    display: flex;
    gap: 40px;
}
