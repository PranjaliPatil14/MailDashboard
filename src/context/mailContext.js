import React, { useContext, useEffect, useReducer } from "react";
import * as PropTypes from "prop-types";
import useLocalStorage, { MAIL } from "../utilities/useLocalStorage";
import { UserContext } from "./userContext";
import {
  mailReducer,
  SET_LOGGED_IN_USER_MAIL,
  SET_VALUES,
} from "./mailReducer";

const initialMailContextValue = {
  mails: [],
  refresh: true,
  loggedInUserMails: [],
  dispatch: () => {},
};

export const MailContext = React.createContext(initialMailContextValue);

const MailContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(mailReducer, initialMailContextValue);
  const { value } = useLocalStorage(MAIL);
  const { loggedInUser } = useContext(UserContext);

  useEffect(() => {
    if (state.refresh) {
      dispatch({ type: SET_VALUES, data: value });
    }
  }, [value, state.refresh]);

  useEffect(() => {
    dispatch({ type: SET_LOGGED_IN_USER_MAIL, data: loggedInUser.id });
  }, [loggedInUser, value, state.mails]);

  useEffect(() => {
    localStorage.setItem(MAIL, JSON.stringify(state.mails));
  }, [state.mails]);

  return (
    <MailContext.Provider value={{ ...state, dispatch }}>
      {children}
    </MailContext.Provider>
  );
};

MailContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default MailContextProvider;
