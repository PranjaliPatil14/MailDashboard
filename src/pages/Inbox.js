import React from "react";
import Header from "../components/header/Header";
import "./inbox.css";

const Inbox = () => {
  return (
    <div className="page">
      <Header />
    </div>
  );
};

export default Inbox;
