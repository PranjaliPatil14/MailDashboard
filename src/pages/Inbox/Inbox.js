import React from "react";
import Header from "../../components/header/Header";
import "./inbox.css";
import Menu from "./menu/Menu";
import Mails from "./mail/Mails";

const Inbox = () => {
  return (
    <div className="page">
      <Header />
      <div className="page-content">
        <Menu />
        <Mails />
      </div>
    </div>
  );
};

export default Inbox;
