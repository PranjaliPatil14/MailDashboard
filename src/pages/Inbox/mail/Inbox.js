import React, { useContext, useState } from "react";
import "./inbox.css";
import MailHeader from "./MailHeader";
import Mail from "./Mail";
import { MailContext } from "../../../context/mailContext";

const Inbox = () => {
  const { loggedInUserMails } = useContext(MailContext);
  const [selectedMails, setSelectedMails] = useState([]);

  const sortedMails = loggedInUserMails.sort((a, b) => {
    return new Date(b.time) - new Date(a.time);
  });

  return (
    <div className="inbox-mails">
      <MailHeader selectedMails={selectedMails} />
      <div className="mails-content">
        {sortedMails.map((mail) => (
          <Mail mail={mail} setSelectedMails={setSelectedMails} key={mail.id} />
        ))}
      </div>
    </div>
  );
};

export default Inbox;
