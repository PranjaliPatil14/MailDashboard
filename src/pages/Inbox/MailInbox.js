import React, { useContext, useEffect, useState } from "react";
import Header from "../../components/header/Header";
import "./mailInbox.css";
import Menu from "./menu/Menu";
import Inbox from "./mail/Inbox";
import { UserContext } from "../../context/userContext";
import ComposeEmail from "../ComposeEmail/ComposeEmail";

const MailInbox = () => {
  const { setLoggedInUser } = useContext(UserContext);
  const [showCompose, setShowCompose] = useState(false);

  useEffect(() => {
    setLoggedInUser({
      id: 2,
      name: "abc",
      role: "Art Director",
      profileImage: "profile-image.png",
    });
  }, [setLoggedInUser]);
  return (
    <div className="page">
      {showCompose && <ComposeEmail setShowCompose={setShowCompose} />}
      <Header />
      <div className="page-content">
        <Menu setShowCompose={setShowCompose} />
        <Inbox />
      </div>
    </div>
  );
};

export default MailInbox;
