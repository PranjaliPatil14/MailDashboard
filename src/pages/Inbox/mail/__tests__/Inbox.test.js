import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Inbox from "../Inbox";
import { MailContext } from "../../../../context/mailContext";
import mails from "../../../../mockData/mails.json";
import mockUsers from "../../../../mockData/users.json";
import { UserContext } from "../../../../context/userContext";
import { CHANGE_IMPORTANCE } from "../../../../context/mailReducer";

describe("Inbox", () => {
  const loggedInUserMails = mails.splice(0, 2);
  const users = mockUsers;
  const dispatch = jest.fn();
  test("should render sorted Mails", () => {
    const { asFragment, container } = render(
      <UserContext.Provider value={{ users }}>
        <MailContext.Provider value={{ loggedInUserMails }}>
          <Inbox />
        </MailContext.Provider>
      </UserContext.Provider>
    );

    expect(
      asFragment(container.querySelector(".mails-content"))
    ).toMatchSnapshot();
  });

  test("should mark mail as important", () => {
    const { getAllByRole, getByTestId } = render(
      <UserContext.Provider value={{ users }}>
        <MailContext.Provider value={{ loggedInUserMails, dispatch }}>
          <Inbox />
        </MailContext.Provider>
      </UserContext.Provider>
    );

    fireEvent.click(getAllByRole("checkbox")[0]);

    fireEvent.click(getByTestId(CHANGE_IMPORTANCE));

    expect(dispatch).toHaveBeenCalledWith({
      data: [2],
      type: "CHANGE_IMPORTANCE",
    });
  });
});
