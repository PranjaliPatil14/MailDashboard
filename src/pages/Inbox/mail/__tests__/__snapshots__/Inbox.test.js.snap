// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inbox should render sorted Mails 1`] = `
<DocumentFragment>
  <div
    class="inbox-mails"
  >
    <div
      class="mails-header"
    >
      <div
        class="mails-header-title"
      >
        <h2>
          Inbox (2)
        </h2>
        <div
          class="mail-search"
        >
          <input
            placeholder="Search email"
            type="search"
          />
          <button
            class="button "
            data-testid=""
            type="submit"
          >
            Search
          </button>
        </div>
      </div>
      <div
        class="mail-operations"
      >
        <div
          class="mail-actions"
        >
          <button
            class="button mail-action"
            data-testid="REFRESH_MAILS"
            type="submit"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="23 4 23 10 17 10"
              />
              <polyline
                points="1 20 1 14 7 14"
              />
              <path
                d="M3.51 9a9 9 0 0 1 14.85-3.36L23 10M1 14l4.64 4.36A9 9 0 0 0 20.49 15"
              />
            </svg>
             Refresh
          </button>
          <button
            class="button mail-action"
            data-testid="CHANGE_READ_STATUS"
            type="submit"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1 12s4-8 11-8 11 8 11 8-4 8-11 8-11-8-11-8z"
              />
              <circle
                cx="12"
                cy="12"
                r="3"
              />
            </svg>
          </button>
          <button
            class="button mail-action"
            data-testid="CHANGE_IMPORTANCE"
            type="submit"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="10"
              />
              <line
                x1="12"
                x2="12"
                y1="8"
                y2="12"
              />
              <line
                x1="12"
                x2="12.01"
                y1="16"
                y2="16"
              />
            </svg>
          </button>
          <button
            class="button mail-action"
            data-testid="MARK_DELETED"
            type="submit"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="3 6 5 6 21 6"
              />
              <path
                d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
              />
              <line
                x1="10"
                x2="10"
                y1="11"
                y2="17"
              />
              <line
                x1="14"
                x2="14"
                y1="11"
                y2="17"
              />
            </svg>
          </button>
        </div>
        <div
          class="mail-navigation"
        >
          <button
            class="button mail-action"
            data-testid=""
            type="submit"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="19"
                x2="5"
                y1="12"
                y2="12"
              />
              <polyline
                points="12 19 5 12 12 5"
              />
            </svg>
          </button>
          <button
            class="button mail-action"
            data-testid=""
            type="submit"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="5"
                x2="19"
                y1="12"
                y2="12"
              />
              <polyline
                points="12 5 19 12 12 19"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
    <div
      class="mails-content"
    >
      <div
        class="mail"
      >
        <input
          type="checkbox"
        />
        <div
          class="mail--important"
        />
        <p
          class="mail--from"
        >
          pqr
        </p>
        <div
          class="mail--tags"
        >
          <span>
            Jobs
          </span>
        </div>
        <p
          class="mail--body"
        >
           Quisque non lectus tincidunt, consequat sapien eu, rutrum nisl.
        </p>
        <span
          class="mail--attachment"
        />
        <span
          class="mail-deliveryTime"
        >
          Jan 5
        </span>
      </div>
      <div
        class="mail"
      >
        <input
          type="checkbox"
        />
        <div
          class="mail--important"
        />
        <p
          class="mail--from"
        >
          abc
        </p>
        <div
          class="mail--tags"
        >
          <span>
            Development
          </span>
        </div>
        <p
          class="mail--body"
        >
           Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla eget rhoncus metus.
        </p>
        <span
          class="mail--attachment"
        />
        <span
          class="mail-deliveryTime"
        >
          Jan 2
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
